src/serialbox-fortran/utils_ppser_kbuff.f90

```diff
@@ -0,0 +1,684 @@
+  INTEGER, INTENT(OUT)                    :: kbuff_id
+
+  ! local vars
+  INTEGER :: call_index = 0
+  INTEGER :: i
+      
+  IF (debug) THEN
+    WRITE(0,*) 'DEBUG setup_buffer: savepoint=', TRIM(savepoint%savepoint_name)
+    WRITE(0,*) 'DEBUG setup_buffer: fieldname=', TRIM(fieldname)
+    WRITE(0,*) 'DEBUG setup_buffer: k=', k
+    WRITE(0,*) 'DEBUG setup_buffer: k_size=', k_size
+  END IF
+
+  ! ppser mode numbers do not align with m_serialize constants....
+  IF ( mode /= PPSER_MODE_WRITE ) THEN
+    WRITE(0,*) 'ERROR, can only use kbuffer in write mode'
+    STOP
+  END IF
+ 
+  ! initialize if this is the first call
+  IF ( first_call ) THEN
+    CALL init_kbuff()
+  END IF
+
+  ! find ID if it already exists
+  CALL find_kbuff_id(fieldname, savepoint, k, kbuff_id, call_index)
+  IF (debug) THEN
+    WRITE(0,*) 'DEBUG fs_write_kbuff_float_3d_r8: find kbuff_id=', kbuff_id
+  END IF
+
+  ! check if a buffer slot was found
+  IF ( kbuff_id == 0 ) THEN
+    ! no, so create a new buffer
+    CALL get_free_kbuff_id(kbuff_id)
+    IF (debug) THEN
+      WRITE(0,*) 'DEBUG fs_write_kbuff_float_3d: kbuff_id=', kbuff_id
+    END IF
+    CALL create_kbuff(kbuff_id, serializer, savepoint, fieldname, field_type, &
+                     field_nx, field_ny, k_size, call_index, minushalos, plushalos)
+  ELSE
+    ! yes, so check for consistency of current request with stored metadata
+    CALL check_kbuff(kbuff_id, serializer, savepoint, fieldname, field_type, &
+                     field_nx, field_ny, k_size, k, minushalos, plushalos)
+  END IF
+
+ END SUBROUTINE setup_buffer
+
+!============================================================================
+
+
+! create a new buffer (allocate memory, store metadata)
+SUBROUTINE create_kbuff(kbuff_id, serializer, savepoint, fieldname, field_type, &
+                       dim_i, dim_j, dim_k, call_index, minushalos, plushalos)
+  IMPLICIT NONE
+
+  INTEGER, INTENT(IN)                     :: kbuff_id
+  TYPE(t_serializer), TARGET, INTENT(IN)  :: serializer
+  TYPE(t_savepoint), TARGET, INTENT(IN)   :: savepoint
+  CHARACTER(LEN=*), INTENT(IN)            :: fieldname
+  INTEGER, INTENT(IN)                     :: field_type
+  INTEGER, INTENT(IN)                     :: dim_i, dim_j, dim_k
+  INTEGER, INTENT(IN)                     :: call_index
+  INTEGER, INTENT(IN), OPTIONAL           :: minushalos(3), plushalos(3)
+
+  ! debug information
+  IF (debug) THEN
+    WRITE(0,*) 'DEBUG create_kbuff: kbuff_id=', kbuff_id
+    WRITE(0,*) 'DEBUG create_kbuff: savepoint=', TRIM(savepoint%savepoint_name)
+    WRITE(0,*) 'DEBUG create_kbuff: fieldname=', TRIM(fieldname)
+    WRITE(0,*) 'DEBUG create_kbuff: field_type=', field_type
+    WRITE(0,*) 'DEBUG create_kbuff: dim_i,dim_j,dim_k=', dim_i, dim_j, dim_k
+  END IF
+
+  ! security check
+  IF (kbuff_id < 1 .OR. kbuff_id > max_kbuff) THEN
+    WRITE(0,*) 'ERROR in utils_ppser_kbuff: illegal kbuff_id encountered'
+    STOP
```

Comment:
Are you sure you want a stop in utility routines ? Alternative would be to return and give error code (but then code needs to be checked on the user side)

Reply:
Good point. Unfortunately, there's nobody to catch the error, and connecting pp_ser.py with COSMO's error handler is not the way to go. I looked at `m_serialize.f90` how this is handled and adopted the `STOP` strategy from there. I think there is not good alternative.